### PR TITLE
Fix and enable balo tests and disallow invocation expr in lhs of assignment

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Fix and enable balo tests and disallow invocation expr in lhs of assignment and disallow versioned imports

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/18035

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/annotation/AnnotationTests.java [hunk 1]
```diff
@@ -24,11 +24,13 @@
 import org.testng.Assert;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Ignore;
 import org.testng.annotations.Test;
 
 /**
  * Test cases for reading annotations.
  */
+@Ignore
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/annotation/AnnotationTests.java]: Related to https://github.com/ballerina-platform/ballerina-lang/issues/17999

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/globalvar/ModuleLevelVariableTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/balo/globalvar/ModuleLevelVariableTest.java]: Wrong year in the header.

File: tests/jballerina-unit-test/src/test/resources/test-src/balo/test_balo/constant/map-literal-constant-panic.bal [hunk 3]
```diff
@@ -103,35 +95,35 @@ function getByteMap() returns map<byte> {
 // -----------------------------------------------------------
 
 function updateConstantByteMapValueInArrayWithExistingKey() {
-    foo:getBYTEA1()[0].bytem11k = 8; // panic
+    foo:getBYTEA1()[0]["bytem11k"]   = 8; // panic
 }
 
 function updateConstantByteMapValueInArrayWithNewKey() {
-    foo:getBYTEA1()[0].newKey = 8; // panic
+    foo:getBYTEA1()[0]["newKey"] = 8; // panic
```
Comment (hasithaa) [inline tests/jballerina-unit-test/src/test/resources/test-src/balo/test_balo/constant/map-literal-constant-panic.bal]: we shouldn't support this syntax.  A function call is not allowed in LHS. Please create an issue, This is an L1. 
```
lvexpr :=
   variable-reference-lvexpr
   | field-access-lvexpr
   | member-access-lvexpr
variable-reference-lvexpr := variable-reference
field-access-lvexpr := lvexpr . field-name
member-access-lvexpr := lvexpr [ expression ]
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 4]
```diff
@@ -2423,13 +2424,24 @@ void addAssignmentStatement(DiagnosticPos pos, Set<Whitespace> ws) {
...
         assignmentNode.setExpression(rExprNode);
         assignmentNode.pos = pos;
         assignmentNode.addWS(ws);
         assignmentNode.varRef = ((BLangVariableReference) lExprNode);
         addStmtToCurrentBlock(assignmentNode);
     }
 
+    private void validateLvexpr(ExpressionNode lExprNode, DiagnosticCode errorCode) {
+        if (lExprNode.getKind() == NodeKind.INVOCATION) {
+            dlog.error(((BLangInvocation) lExprNode).pos, errorCode);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Should we return here?

Also, can't we validate for reference, field access and index based access specifically?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java [hunk 5]
```diff
@@ -2467,7 +2479,9 @@ void addCompoundAssignmentStatement(DiagnosticPos pos, Set<Whitespace> ws, Strin
         BLangCompoundAssignment assignmentNode =
                 (BLangCompoundAssignment) TreeBuilder.createCompoundAssignmentNode();
         assignmentNode.setExpression(exprNodeStack.pop());
-        assignmentNode.setVariable((BLangVariableReference) exprNodeStack.pop());
+
+        BLangVariableReference varref = (BLangVariableReference) exprNodeStack.pop();
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Do we need this change?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 6]
```diff
@@ -1146,3 +1146,9 @@ error.type.param.outside.lang.module=\
 
 error.versioned.import.not.supported=\
   versioned import is not supported
+
+error.invalid.lvalue.lhs.of.assignment=\
+  invocations are not supported in lhs of an assignment
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: how about
"invocations are not supported on the left hand side of an assignment"

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 7]
```diff
@@ -1146,3 +1146,9 @@ error.type.param.outside.lang.module=\
...
+
+error.invalid.lvalue.lhs.of.compound.assignment=\
+  invocations are not supported in lhs of an compound assignment
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties]: "invocations are not supported on the left hand side of a compound assignment"

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java [hunk 8]
```diff
@@ -83,9 +83,19 @@ public void testNegativeCases() {
         validateError(negativeResult, i++, "undefined field 'y' in record 'E'", 75, 5);
         validateError(negativeResult, i++, "invalid operation: type 'map<int>?' does not support member access for " +
                 "assignment", 78, 5);
-        validateError(negativeResult, i, "invalid operation: type 'E?' does not support field access", 79, 5);
+        validateError(negativeResult, i++, "invalid operation: type 'E?' does not support field access", 79, 5);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java]: do we need this change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 5 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: This was a left over from compound assignment validation, will remove this.

File: tests/jballerina-unit-test/src/test/resources/test-src/balo/test_balo/constant/map-literal-constant-panic.bal (hunk 3 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/resources/test-src/balo/test_balo/constant/map-literal-constant-panic.bal]: Fix and validation added to prohibit invocation expr in lhs of an assignment. Validation is already there for compound assignment.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 4 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Didn't understand the comment here.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: What I meant was if we can check if it is 
variable-reference-lvexpr
   | field-access-lvexpr
   | member-access-lvexpr
specifically. 

Maybe we can revisit this later.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java [hunk 9]
```diff
@@ -83,9 +83,19 @@ public void testNegativeCases() {
...
     }
 
+    @Test
+    public void testNegativeLvexpr() {
+        CompileResult negative = BCompileUtil.compile("test-src/statements/assign/lvexpr_negative.bal");
+        int i = 0;
+        validateError(negative, i++, "invocations are not supported in lhs of an assignment", 26, 5);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java]: shouldn't this be updated to reflect the change to the error message?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java (hunk 4 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangPackageBuilder.java]: Ah, only invocation expr seems to leak into lvexpr that is not  allowed, yes let's revisit this later.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java (hunk 9 above)
Comment (rdhananjaya) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/assign/LValueTest.java]: Yes I forgot

**Pull request merged**